write.go

```diff
@@ -142,6 +143,46 @@ func (mw *msgWriter) putFlateWriter() {
 	}
 }
 
+// writeFull compresses and writes p as a single frame.
+func (mw *msgWriter) writeFull(ctx context.Context, p []byte) (int, error) {
+	mw.ensureFlate()
+
+	buf := bpool.Get()
+	defer bpool.Put(buf)
+
+	// Buffer compressed output so we can write as a single frame instead
+	// of chunked frames.
+	origWriter := mw.trimWriter.w
+	mw.trimWriter.w = buf
+	defer func() {
+		mw.trimWriter.w = origWriter
+	}()
+
+	_, err := mw.flateWriter.Write(p)
```

Comment:
Bug: Using flateWriter without holding writeMu (Write implementation below gives a clue that this is necessary). If the connection closes right now, flateWriter may be nil.